src/parser/core/modules/AlwaysBeCastingTable.tsx

```diff
@@ -0,0 +1,304 @@
+	 * @param {Action['id']} actionID ending action. Optional only in the event of the end of the fight
+	 */
+	private checkAndSave(endTime: number, eventAction?: Events['action']) {
+		//no window no problem
+		const tracker = this.noCastWindows
+		if (tracker.current === undefined) { return }
+
+		//return and reset if no violation
+		if (!(endTime - tracker.current.start > tracker.current.recastTime + tracker.current.casterTax + GCD_JITTER_OFFSET
+				|| (tracker.current.interruptedActions !== undefined && tracker.current.interruptedActions?.length !== 0))) {
+			tracker.current = undefined
+			return
+		}
+
+		// Close the window
+		tracker.current.stop = endTime
+		if (eventAction !== undefined) { tracker.current.stopAction = eventAction }
+		tracker.history.push(tracker.current)
+		tracker.current = undefined
+	}
+
+	/**
+	 * Estimate recast based on GCD Length and base recast rounded up to 2 decimal places
+	 * @param {Action['id']} action - desired action for recast
+	 * @returns {number} - estimated recast for desired action, if no cast mentioned, returns base gcdLength
+	 */
+	private getRecast(action: Action['id']): number {
+		const gcdLength: number = this.gcd.getDuration()
+		const actionAttribute: Action['speedAttribute'] | undefined = this.data.getAction(action)?.speedAttribute
+		const actionGCDRecastTime: number = this.data.getAction(action)?.gcdRecast ?? 0
+		//TODO need to consider swiftcast since gcd could be less than recast (see blm)
+		/**
+		 * if there is no recast time, take base estimated gcd length
+		 * if action speed attribute is undefined, assume recast time is always raw e.g. PCT canvas
+		 * if there is a recast time impacted by speed attribute, estimate recast by base recast (2.5s) and actual gcd length.
+		 * In this case, it's rounded up at 2 decimal places to give benefit of doubt
+		 */
+		const unroundedActionRecast: number
+			= actionGCDRecastTime === 0 ? gcdLength
+				: actionAttribute === undefined ? actionGCDRecastTime
+					: gcdLength / BASE_RECAST * actionGCDRecastTime
+		const actionRecast: number = Math.ceil(unroundedActionRecast * 100)/100
+		return actionRecast
+	}
+
+	private onComplete(event: Events['complete']) {
+		//finish up
+		this.checkAndSave(event.timestamp)
+
+		// Filter out periods where you got stunned, boss is untargetable, etc, or windows with negative durations
+		this.noCastWindows.history = this.noCastWindows.history.filter(windows => {
+			const duration = this.downtime.getDowntime(
+				windows.start,
+				windows.stop ?? windows.start,
+			)
+			return duration === 0
+				&& (
+					((windows.stop ?? windows.start) - windows.start > windows.recastTime + windows.casterTax + GCD_JITTER_OFFSET)
+					|| (windows.interruptedActions !== undefined && windows.interruptedActions?.length !== 0)
+				)
+		})
+	}
+
+	override output() {
+		if (this.noCastWindows.history.length === 0) { return <><Trans id="core.always-be-casting-table.no-violations">Nothing to note.</Trans> 😉</> }
```

Comment:
I'd probably just return null/undefined if there aren't any issues so the table doesn't appear in the output at all. That's more to the pattern we follow for other modules where no errors need displaying.

Reply:
updated!